### ISSUE TITLE
Allow null in secondary index

### ISSUE THREAD

**Issue by neumino** [rethinkdb/rethinkdb#1032]
I expect

```
r.db("foo").table("bar").filter( r.row("field").eq( value ) )
```

to return exactly the same results as

```
r.db("foo").table("bar").getAll( value, { index: "field" })

```

Whatever the value of `value` is, including null. I could understand that we don't want to index array/object, but we should index null.

**Comment by stuartpb** [comment]
> If people want to exclude a row from an index, they can still do so by producing a normal error (e.g. with r.error).

Per my comment at https://github.com/rethinkdb/docs/issues/731#issuecomment-95984237, I would rather there be a separate, discrete symbol for the purposes of omitting a value from a secondary index that isn't conflated with runtime / consistency errors - basically a non-serializable value that, like `r.minval` and `r.maxval` before it, stands in for the original meaning of `null`, explicitly, distinguishing it from a literal `null` value.

**Comment by danielmewes** [comment]
@stuartpb The idea behind allowing `null` in secondary indexes is that we want an indexed `orderBy({index: ...})` or a `getAll` to behave the same way as its un-indexed equivalents `orderBy` and `filter({a: x})`. That will allow us to automatically use indexes if they are available in the future.

We could allow such a special value `r.dontIndex` as a return value of an index function, as long as we don't allow it to appear in any other result or document.

At the moment this isn't necessary, because we currently swallow errors that happen in secondary index functions anyway, so the effect of using `r.error` is the same as what that special value would do.

Us not passing up errors in secondary index functions has two reasons at least:
* If you create an index over a field that only exists in some documents, you usually don't want the index creation to fail if there's some document in the table that doesn't have that field.
* If you update or insert an document such that its new value causes an error in the secondary index function, we currently don't have a great way for reporting that error in the query result. We cannot make the full query fail because a) this would probably be undesired and b) the way secondary index updates are implemented at the time they are evaluated the update might already have been applied to other indexes (including the main tree). We could probably change the latter part by evaluating sindex functions *before* writing anything, but I'm not sure if we want to.

We could add a new `sindex_errors` field to the result of write operations where we aggregate all errors that occurred while evaluating the secondary index functions on the changed object. This might be useful for debugging. I'll open a new issue to discuss that.

**Comment by danielmewes** [comment]
@stuartpb See https://github.com/rethinkdb/rethinkdb/issues/4118 for a discussion on how to handle secondary index errors.

**Comment by danielmewes** [comment]
To clarify: I like the idea of having a special value that can be returned from an index function for omitting a document from the index. It would be a much cleaner replacement for using the error trick. I think it's somewhat orthogonal to the `null` question though.

**Comment by coffeemug** [comment]
`r.skip`? :smile:

**Comment by thelinuxlich** [comment]
r.skip feels like pagination syntax

**Comment by stuartpb** [comment]
`r.skip` is the name I was originally going to suggest, but I was thinking it might be better to reuse another value like `r.absent` so there aren't so many single-purpose constants hanging around under `r`. (Also, as @thelinuxlich points out, `r.skip` sounds like a name that might be better suited to another purpose down the line.)

Another possible name: `r.unindexed`

**Comment by stuartpb** [comment]
Anyway, as @danielmewes said, this is orthogonal (though related) to allowing `null` values in secondary indices, so I'll open another issue for it.

**Comment by coffenbacher** [comment]
Is r.absent or r.unindexed still under consideration? This would be really useful in some situations.

**Comment by danielmewes** [comment]
@coffenbacher Yes, though right now you can just return `null` which has the same effect.

**Comment by coffenbacher** [comment]
Does it? We're basically trying to run .getAll(null,index='secondary') to select records that do not have a value set for the secondary index. (This is equivalent to one of the scenarios @mlucy mentioned in an early comment). 

Is there a workaround for that?

**Comment by danielmewes** [comment]
@coffenbacher At the moment `null` values don't get indexed (and this issue is about changing that in the future).
There are multiple work-arounds. One is to map a missing value to something other than `null`, that is not otherwise used in the indexed field(s). For example to `false`:
```js
r.table(...).indexCreate("secondary", r.row("secondary").default(false))
```

Then you can query by `.getAll(false, {index: "secondary"})`.

**Comment by coffenbacher** [comment]
Thanks @danielmewes, that solution worked perfectly for us.

**Comment by mike-marcacci** [comment]
I'd like to add to this proposal. As far as I currently understand, the plan is to:

- start indexing `null` values
- ignore rows that return an "undefined" type using `r.absent()` or the existing `r.literal()`

I'd like to also reconsider the behavior of an `r.error()` (or an implicit error, like operating on the wrong type or not specifying a default) when returned by an index function. I'd propose that a write operation fail if *any* index returns an error. Likewise, I would propose that creating an index should fail if *any* documents return such an error.

This would ensure our indices are expecting the correct document structures, and aren't silently missing documents that we would expect to have indexed. This would also have the effect of validating every document in the table.

To ensure backwards-compatibility and ease the process of schema migrations, `indexCreate` could receive a boolean option of `strict` to enable this behavior.

Piggy-backing on the same logic, rethinkdb could introduce the concept of "validators" which would function similarly to "strict" indices, but wouldn't actually create the b-trees. This could essentially accomplish #5494, #5879, and #4210, as any json-schema or Joi schema could be compiled into a rql expression and stored as a "validator" expression in the database.
